### PR TITLE
Add Compressed Sparse Attention (CSA) module for Deepseek v4 Pro/Flash

### PR DESCRIPTION
**Pull request by ddjekicTT**
https://github.com/tenstorrent/tt-metal/issues/54456

### Summary
Compressed Sparse Attention is V4's long-context attention layer: instead of HCA's compress-every-128-tokens, it compresses every 4 tokens with an overlap transform (each window pools its own K/V together with its predecessor window's first half), then runs a Lightning Indexer over the resulting compressed entries so each query attends to at most index_topk of them, on top of a sliding window over raw tokens and a per-head attention sink. Assembling it needs three pieces that did not exist: 

1. A ratio-4 overlap compressor that works on an SP-sharded sequence and can hand its boundary state both to the next SP chip and to the next prefill chunk
2. An indexer that scores compressed entries rather than tokens and keeps its own persistent compressed-key cache across chunks, 
3. The CSA block itself, which appends to the compressed cache, turns the indexer's picks into an attention mask, and reuses V4's existing attention core. 

(All three are included  here, along with the device ops they need (csa_compressor, compressor_state_exchange, and ratio-awareness in indexer_score) and a shared TtV4AttentionBase extracted from HCA.)

### How compression is done
CSA pooling at compress rate 4 with a one-window overlap. Every group of 4 tokens ("window") produces one pooled row, but the softmax pool runs over 8 candidates: 4 from the previous window and 4 from the current one.

That's why kv and gate are [1, 1, S_local, 2*head_dim] — the left half of the width is the Ca projection (contributed to the next window) and the right half is Cb (contributed to its own window). head_dim is derived as width / 2.

The torch reference at models/demos/deepseek_v3_d_p/tests/op_unit_tests/test_csa_compressor.py states it directly:

```
test_csa_compressor.py
Lines 63-87
def _compress_local(kv, gate, position_bias, predecessor_kv, predecessor_score, start_position, head_dim):
    # ... for each window:
        overlap_kv = torch.cat([previous_ca_kv, current_kv[..., head_dim:]], dim=2)
        overlap_score = torch.cat([previous_ca_score, current_gate[..., head_dim:]], dim=2)
        weights = overlap_score.softmax(dim=2, dtype=torch.float32).to(overlap_kv.dtype)
        pooled[:, :, window] = (overlap_kv * weights).sum(dim=2)
```
**The crucial structural fact:** the softmax is over the 8 candidates independently per feature column. Scores are full head_dim-wide, not scalars. That is what makes the whole thing expressible as elementwise tile math, and the kernel design leans entirely on it.

### The overlap state, and why it exists
Window 0 on a chip has no previous window locally — its Ca candidates live on the previous SP chip (or, for chunked prefill, in the previous chunk). The op carries that across in a 64-row state slab per chip, shaped [1, 1, 64, head_dim], one for KV and one for scores. The layout is documented host-side in models/demos/deepseek_v3_d_p/tt/mla/compressor.py:

```
compressor.py
Lines 48-52
    slot = position % compress_rate
    parity = (position // compress_rate) & 1
    ca_row = (parity ^ 1) * ttnn.TILE_SIZE + slot
    cb_row = parity * ttnn.TILE_SIZE + compress_rate + slot
```

The slab is two 32-row tiles indexed by window parity. A window writes its Ca into the opposite-parity tile — which is exactly where its successor (opposite parity by construction) looks for a predecessor. So consecutive windows never touch the same tile, and one can be read while the other is being filled. Only rows 0–7 and 32–39 are ever live; the other 48 rows are pure pass-through from the base state. The 64-row shape is chosen because it's what Blaze decode consumes, so prefill and decode share it.

image

### Top level: `ttnn.experimental.deepseek_prefill.csa_compressor()`

image

### Compression (per output tile)
image

### Compressor stacked perf on bh mesh 2x2 (640 local seq len)
image

#### NOTE: The current implementation is using dense sdpa with a mask (functional only) the plan is to use sparse sdpa on a joint sw+compressed KV which depends on [this PR](https://github.com/tenstorrent/tt-metal/pull/55755) adding attention sink to sparse sdpa.

image

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ddjekic/ds_v4_csa)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ddjekic/ds_v4_csa)

### CODE REVIEW

**Comment by pavlejosipovic** [comment]
@ddjekicTT can you break this up into stacked PRs. One PR perf op?

File: ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp [hunk 1]
```diff
@@ -42,9 +42,14 @@ void bind_indexer_score(nb::module_& mod) {
...
             chunk_start_idx: absolute global position of rank 0's query row 0
                 (rank 0 = lowest seq_shard_axes[0] (SP) coord; causality: key t
-                visible to query s iff t <= chunk_start + s). OMIT on a mesh ->
-                deduced as T - sp_ring*Sq (sp_ring = mesh extent along the SP axis,
+                visible to query s iff t < floor((chunk_start + s + 1) /
+                key_compression_ratio)). OMIT on a mesh -> deduced from the compressed
+                K length as T*key_compression_ratio - sp_ring*Sq (sp_ring = mesh extent along the SP axis,
                 whole mesh if unset). Single device: set to history + rank*Sq per rank.
+            key_compression_ratio: query tokens represented by one K-cache row. 1
+                (default) preserves DeepSeek-V3.2/GLM behavior; 4 implements DeepSeek
```
Comment (dlukicTT) [inline ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp]: [nit] Looks like Claude's historical comment. Consider rephrasing :)

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        runtime_args.reserve(11);
+        runtime_args.push_back(args.kv.buffer());
+        runtime_args.push_back(args.gate.buffer());
+        runtime_args.push_back(args.position_bias.buffer());
+        runtime_args.push_back(args.base_kv_state.buffer());
+        runtime_args.push_back(args.base_score_state.buffer());
+        runtime_args.push_back(outputs[0].buffer());
+        runtime_args.push_back(outputs[1].buffer());
+        runtime_args.push_back(local_valid);
+        runtime_args.push_back(absolute_start);
```
Comment (dlukicTT) [inline ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp:111]: Can we put these args in a struct which owns serialization to this vector and costruction from the bytearray on kernel size?
I know it's doable with CT args, I'm guessing it's doable with RT args as well, probably in even less convoluted way :)

Comment (dlukicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp:111]: Also, there are now some utils for passing these args, like str->int dictionaries so at least the args are named and not fetched from cryptic byte-array positions. So you can use that as well, although struct looks like a better solution.

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+inline uint16_t float_to_bf16_rne(float value) {
+    uint32_t bits;
+    std::memcpy(&bits, &value, sizeof(bits));
+    bits += 0x7FFFu + ((bits >> 16) & 1u);
+    return static_cast<uint16_t>(bits >> 16);
+}
+
+// One tile row is two runs of 16 contiguous elements, one per face column, and both start on a
+// 32-byte boundary. Moving them as words keeps the copy to 16 stores per row.
+inline void copy_tile_row(
```
Comment (dlukicTT) [inline ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp:48]: Consider pulling out in a common hpp file

**Review by mtairum — APPROVED** [review]
Approving from the test pipelines side.

**Comment by jmitrovicTT** [comment]
CI considerations:

Pls add Bh E2E and Blaze runs links to PR description, 
Also all tests from `models/demos/deepseek_v3_d_p/tests/pcc/` will run in BH E2E pipeline - check if some should be uncollected

In Blaze CI you are running `3chunk-varying` case which is 512+1024+1024 so indexer doesn't do much there, consider increasing a chunk size and actual_seq_len in this test, also It will be nice having same test for PRO in Blaze

Nitpick: you can add a `csa` tag to `.github/workflows/blaze-models-prefill-tests.yaml`

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/csa_compressor.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        gate,
+        position_bias,
+        initial_kv_state,
+        initial_score_state,
+        seq_len_actual,
+        first_token_position,
+        cluster_axis);
+
+    auto [predecessor_kv, predecessor_score] = compressor_state_exchange::compressor_state_exchange(
+        local_state[0], local_state[1], initial_kv_state, initial_score_state, cluster_axis, topology);
```
[the remark this answers is not in the document]
Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/csa_compressor.cpp:32]: Valid point.. Fixed it.

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/compressor_state_exchange/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        TT::Metalium
+    PUBLIC
+        TTNN::Core
+        TTNN::Ops::CCL
+        TTNN::Ops::DataMovement
+        TTNN::Ops::PointToPoint
+)
+
+install(TARGETS ttnn_op_experimental_deepseek_prefill_compressor_state_exchange LIBRARY COMPONENT tar)
+install(TARGETS ttnn_op_experimental_deepseek_prefill_compressor_state_exchange FILE_SET api COMPONENT ttnn-dev)
```
[the remark this answers is not in the document]
Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/compressor_state_exchange/CMakeLists.txt:50]: Valid point.. fixed it

**Comment by ddjekicTT** [comment]
> @ddjekicTT can you break this up into stacked PRs. One PR perf op?
@pavlejosipovic 

I really thought about it.. But there are only two ops, and one is basically the halo exchange for the other one.. It would be hard reviewing them out of context.

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp (hunk 3 above)
Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp]: Agreed

**Comment by ddjekicTT** [comment]
> CI considerations:
> 
> Pls add Bh E2E and Blaze runs links to PR description, Also all tests from `models/demos/deepseek_v3_d_p/tests/pcc/` will run in BH E2E pipeline - check if some should be uncollected
> 
> In Blaze CI you are running `3chunk-varying` case which is 512+1024+1024 so indexer doesn't do much there, consider increasing a chunk size and actual_seq_len in this test, also It will be nice having same test for PRO in Blaze
> 
> Nitpick: you can add a `csa` tag to `.github/workflows/blaze-models-prefill-tests.yaml`
@jmitrovicTT 
Thanks. I added a 3chunk-ragged scenario [5120, 5120, 3000] for 8x4 mesh grid, and a recall check for the indexer.

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp (hunk 3 above)
Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/kernels/csa_state_update.cpp]: Done.

Moved the duplicated tile constants and helpers into csa_compressor_kernel_utils.hpp

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp (hunk 2 above)
Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp]: Good idea. I switched to using a named struct, but I realized I could lose the Buffer* typing.. I’ve instead added typed host-side structs that own serialization into RTArgList

Comment (ddjekicTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/csa_compressor/device/csa_compressor_program_factory.cpp]: There is also something called `blaze_named_args` but it is experimental and scheduled for removal so I opted out

**Comment by pavlejosipovic** [comment]
> > @ddjekicTT can you break this up into stacked PRs. One PR perf op?
> > @pavlejosipovic
> 
> I really thought about it.. But there are only two ops, and one is basically the halo exchange for the other one.. It would be hard reviewing them out of context.

that what stacked PRs are for you get full context, and you get to review only piece you are the code owner, and get that through CI as one thing

**Comment by ddjekicTT** [comment]
> > > @ddjekicTT can you break this up into stacked PRs. One PR perf op?
> > > @pavlejosipovic
> > 
> > 
> > I really thought about it.. But there are only two ops, and one is basically the halo exchange for the other one.. It would be hard reviewing them out of context.
> 
> that what stacked PRs are for you get full context, and you get to review only piece you are the code owner, and get that through CI as one thing

Let's try it that way. I will close this PR in favor of a separate stacked one https://github.com/tenstorrent/tt-metal/pull/57195

**Comment by ddjekicTT** [comment]
This PR is closed in favor of the stacked PR alternative https://github.com/tenstorrent/tt-metal/pull/57195
